### PR TITLE
Support for AutoMod Members

### PR DESCRIPTION
**Pull request by Puncher1**
## Summary

This PR adds support for AutoMod Members. This also adds the new param `mention_raid_protection` in `AutoModTrigger`.
Related PR: https://github.com/discord/discord-api-docs/pull/6040

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Review by bijij — COMMENTED** [review]
I'm very against making a breaking change here, not sure what harm there is in still having `block_message` be the default.

At most I'd say not explicitly passing type should spit back a deprecation warning.

File: docs/api.rst
Comment (kaylynn234) [inline docs/api.rst:3307]: The wording here needs to be updated. It's very unclear what you're trying to say.

File: docs/api.rst
Comment (kaylynn234) [inline docs/api.rst:3308]: "it" is usually used to refer to objects rather than people, so I don't think it's the right fit here. It would probably be better to go with "their" instead?

(and just for the record, "it's" is the incorrect word to use here. "it's" is a shortening of "it is", whereas "its" is the possessive term, and refers to ownership of something)

File: docs/api.rst
Comment (kaylynn234) [inline docs/api.rst:3267]: It might be worth clarifying *which* keywords trigger the rule, since this could seem ambiguous otherwise.

Imagine you're somebody reading the documentation - which keywords trigger the rule? Are these keywords explicitly configured? If so, where are they configured?

**Review by kaylynn234 — COMMENTED** [review]
Implementation aside, I think the documentation could be improved a bit. Nothing major!

**Comment by Puncher1** [comment]
> I'm very against making a breaking change here, not sure what harm there is in still having `block_message` be the default.
> 
> At most I'd say not explicitly passing type should spit back a deprecation warning.

There's just no reason why `block_message` should be the default, but since it would be a breaking change, I agree.

File: discord/automod.py
Comment (Puncher1) [reply discord/automod.py:128]: I would change the first error from
```py
raise ValueError(f'{attribute} can only be passed if type is {action_type.name}.')
```
to
```py
raise ValueError(f'{attribute} can only be passed if type is {action_type.name} or None.')
```
Otherwise I think this is good, thanks!

Comment (Puncher1) [reply discord/automod.py:128]: One thing I noticed in 
```py
if type is None and sum(v is None for v in (channel_id, duration, custom_message)) < 2:
   ^^^^^^^^^^^^
```
`type` is irrelevant since it also applies when `type` is not `None`.

File: discord/automod.py
Comment (Puncher1) [reply discord/automod.py:94]: 1. Yeah true, neither `custom_message` should. 
2. Why? `type` is optional.

File: docs/api.rst
Comment (Puncher1) [reply docs/api.rst:3307]: What about this?
```
Similar to :attr:`timeout` the rule will timeout a user except of having a limit.
```

File: docs/api.rst
Comment (Puncher1) [reply docs/api.rst:3308]: Ah okay, thanks for clarifying. As you may have already noticed, English is not my native language. 😄

File: docs/api.rst
Comment (Puncher1) [reply docs/api.rst:3267]: Yeah, the description is based on the description of `.. attribute:: keyword` which is basically based on the same functionality. Also it's just the enum `AutoModTriggerType` which is needed to create a `AutoModTrigger`. There it's better documented inside a table anyway: 
```py
+-----------------------------------------------+------------------------------------------------+
| :attr:`AutoModRuleTriggerType.member_profile` | :attr:`keyword_filter`, :attr:`regex_patterns`,|
|                                               | :attr:`allow_list`                             |
+-----------------------------------------------+------------------------------------------------+
 ```
 So I think it's fine.

File: docs/api.rst
Comment (kaylynn234) [reply docs/api.rst:3307]: I think the second is a good improvement. It might be good to reorder the sentence a bit for clarity:
```rst
Similar to :attr:`timeout`, except the user will be timed out indefinitely.
```
This also leads nicely into the next sentence, which is a bonus!

File: discord/automod.py
Comment (Puncher1) [reply discord/automod.py]: Initially I wanted it at the end of the constructor, so I forgot to replace `self.type` with `type`. But it's better to have it at the end anyway, so I put it back now.

**Review by Rapptz — CHANGES_REQUESTED** [review]

**Comment by 0xallie** [comment]
Is there anything still holding this PR back?

**Comment by Duued** [comment]
> Is there anything still holding this PR back?

You may have forgotten to re-request review from Rapptz

**Comment by 0xallie** [comment]
Not my PR, just curious about its state.

**Comment by Puncher1** [comment]
It's still not documented on Discord's side, see here: https://github.com/discord/discord-api-docs/pull/6040

**Comment by Puncher1** [comment]
This PR can be merged already. The original PR cannot isn't merged yet due to merge conflicts but the "not released" label has been removed, see this comment: https://github.com/discord/discord-api-docs/pull/6040#issuecomment-1634744229

**Review by Rapptz — CHANGES_REQUESTED** [review]
This PR just needs to change the 2.3 references to 2.4.

**Review by Rapptz — APPROVED** [review]
Thanks!

**Pull request merged**
